robot/base_robot.py

```diff
@@ -151,6 +156,200 @@ def _calculate_reward(self, obs) -> float:
         return 0
 # test commit since github is weird
 
+
+class XArmTaskSpaceEnv(RobotEnv):
```

Comment:
there's a lot of copy and paste in this class. you can use inheritance to your advantage here to clean this up (i.e., override the method, do what you want that's custom, and then call the parent classes version of that function)

then you don't need to redundantly code up functions that exist already also (ex: action_space, _calculate_reward)

you probably want to do somethng like:
```
class XArmEnv(TaskSpaceRobotEnv):
  def __init__(self, xarm_ip: str = '192.168.1.220', **kwargs):
     self.xarm_ip = xarm_ip
     super().__init__(self, **kwargs)
```

Reply:
also this class should probably live in the `xarm_env.py` file so you shouldn't be making any changes to `base_robot.py`